### PR TITLE
UAT insert link to conflicting programs, subplans and courses when a …

### PR DESCRIPTION
**Pull request by joshcbee**
…uniqueness constraint is violated.

The error will only trigger when valid entries have been provided for the constrained fields, as field-level validation is run prior to form level.

The relevant constraints to test are:
- program - unique_together = (("code", "year"), ("name", "year"))
- subplan - unique_together = (("code", "year"), ("year", "name", "planType"),)
- course - unique_together = (("code", "year"),)

image

### CODE REVIEW

File: cassdegrees/ui/forms.py [hunk 1]
```diff
@@ -89,6 +108,45 @@ def clean_units(self):
...
         return data
 
+    # Override clean to return links to existing content if unique_together constraint fails
+    # https://stackoverflow.com/questions/4659360/get-django-object-id-based-on-model-attribute
+    def clean(self):
+        cleaned_data = super().clean()
+
+        # check assignment as keys may not exist in cleaned dictionary if field level validation has failed
+        try:
+            draft_code = cleaned_data['code']
```
Comment (j-selby) [inline cassdegrees/ui/forms.py]: Is there no way to hook into the pre-existing error checking here? No biggie if not.

**Review by j-selby — APPROVED** [review]
Looks good otherwise!

File: cassdegrees/ui/forms.py (hunk 1 above)
Comment (joshcbee) [reply cassdegrees/ui/forms.py]: We're hooking into the pre-existing clean method. But the pre-existing validate_unique() method is a nightmare and goes several layers down (multiple overrides) and from what I can see has no easy place to hook in to actually return a conflicting id.

This method has the benefit that the natural errors raised by validate_unique will specify the conflicting fields with the simple override we've already put in the NON_FIELD_ERRORS, so we can make some generic 'get conflicting record link' methods.

**Review by JackSekoranja — APPROVED** [review]
Looks good and works well, happy for you to commit

**Pull request merged**